noir-projects/noir-protocol-circuits/crates/types/src/utils/arrays/sort_by_counter.nr

```diff
@@ -0,0 +1,160 @@
+        let items = values.map(|value| TestItem { value, counter: value });
+        sort_by_counter_asc(items).map(|item: TestItem| item.value)
+    }
+
+    fn sort_by_values_desc<let N: u32>(values: [u32; N]) -> [u32; N] {
+        let items = values.map(|value| TestItem { value, counter: value });
+        sort_by_counter_desc(items).map(|item: TestItem| item.value)
+    }
+
+    fn compare_test_items_asc(value_1: u32, value_2: u32) -> bool {
+        compare_by_counter_empty_padded_asc(
+            TestItem { value: value_1, counter: value_1 },
+            TestItem { value: value_2, counter: value_2 }
+        )
+    }
+
+    fn compare_test_items_desc(value_1: u32, value_2: u32) -> bool {
+        compare_by_counter_empty_padded_desc(
+            TestItem { value: value_1, counter: value_1 },
+            TestItem { value: value_2, counter: value_2 }
+        )
+    }
+
+    #[test]
+    fn compare_by_counter_empty_padded_asc_expected() {
+        assert_eq(compare_test_items_asc(1, 2), true);
+        assert_eq(compare_test_items_asc(1, 1), false);
+        assert_eq(compare_test_items_asc(2, 1), false);
+        assert_eq(compare_test_items_asc(0, 0), false);
+    }
+
+    #[test]
+    fn compare_by_counter_empty_padded_desc_expected() {
+        assert_eq(compare_test_items_desc(1, 2), false);
+        assert_eq(compare_test_items_desc(1, 1), true);
+        assert_eq(compare_test_items_desc(2, 1), true);
+        assert_eq(compare_test_items_desc(0, 0), false);
+    }
+
+    #[test]
+    fn sort_by_counter_asc_full_non_empty() {
+        let sorted = sort_by_values_asc([4, 2, 1, 3, 5]);
+        assert_eq(sorted, [1, 2, 3, 4, 5]);
+    }
+
+    #[test]
+    fn sort_by_counter_desc_full_non_empty() {
+        let sorted = sort_by_values_desc([4, 2, 1, 3, 5]);
+        assert_eq(sorted, [5, 4, 3, 2, 1]);
+    }
+
+    #[test]
+    fn sort_by_counter_asc_padded_empty() {
+        let sorted = sort_by_values_asc([4, 2, 0, 0, 1, 0, 3, 5]);
+        assert_eq(sorted, [1, 2, 3, 4, 5, 0, 0, 0]);
+    }
+
+    #[test]
+    fn sort_by_counter_desc_padded_empty() {
+        let sorted = sort_by_values_desc([4, 2, 0, 0, 1, 0, 3, 5]);
+        assert_eq(sorted, [5, 4, 3, 2, 1, 0, 0, 0]);
+    }
+
+    #[test]
+    fn sort_by_counter_asc_with_zero_counters() {
+        let original = [
+            TestItem { value: 55, counter: 1 },
+            TestItem { value: 11, counter: 0 },
+            TestItem { value: 33, counter: 2 },
+            TestItem { value: 44, counter: 0 },
+            TestItem { value: 22, counter: 0 },
+            TestItem::empty(),
+            TestItem::empty()
+        ];
+        let expected = [
+            TestItem { value: 11, counter: 0 },
+            TestItem { value: 44, counter: 0 },
+            TestItem { value: 22, counter: 0 },
+            TestItem { value: 55, counter: 1 },
+            TestItem { value: 33, counter: 2 },
+            TestItem::empty(),
+            TestItem::empty()
+        ];
```

Comment:
The previous code would reverse the ordering of items with the same counter. The result would be:
```
[
    TestItem { value: 22, counter: 0 },
    TestItem { value: 44, counter: 0 },
    TestItem { value: 11, counter: 0 },
    TestItem { value: 55, counter: 1 },
    TestItem { value: 33, counter: 2 },
    TestItem::empty(),
    TestItem::empty()
]
```
Which would be wrong when sorting an array that has values emitted from private.